### PR TITLE
Mediapipe body points 33 keypoints

### PR DESCRIPTION
**Pull request by pauchai**
### Checklist

- [ ] I submit my changes into the `develop` branch
- [ ] I have created a changelog fragment 
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues (see [GitHub docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary
  ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
  [cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning),
  [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and
  [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [ ] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

### CODE REVIEW

**Comment by nmanovic** [comment]
@pauchai , thank you for the contribution. Could you please help us to fix the linter issues?
The files will be checked: serverless/tensorflow/mediapipe/nuclio/main.py

```
************* Module main
serverless/tensorflow/mediapipe/nuclio/main.py:92:0: W0311: Bad indentation. Found 28 spaces, expected 16 (bad-indentation)
serverless/tensorflow/mediapipe/nuclio/main.py:1:0: W0611: Unused import cv2 (unused-import)
serverless/tensorflow/mediapipe/nuclio/main.py:5:0: W0611: Unused matplotlib.pyplot imported as plt (unused-import)
```

**Comment by nmanovic** [comment]
@pauchai , could you please add a line into CHANGELOG.md? Please read in CHANGELOG how to do that.

File: serverless/tensorflow/mediapipe/nuclio/function.yaml [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  namespace: cvat
+  annotations:
+    name: Mediapipe human pose estimation
+    type: detector
+    framework: tensorflow
+    spec: |
+      [
+        {
+          "name" : "body",
+          "type" : "skeleton",
```
Comment (nmanovic) [inline serverless/tensorflow/mediapipe/nuclio/function.yaml:13]: Could you please add "svg" as well? Please look here: https://github.com/opencv/cvat/blob/develop/serverless/pytorch/mmpose/hrnet32/nuclio/function.yaml. You can contract it in CVAT UI and download: https://opencv.github.io/cvat/docs/manual/advanced/annotation-with-skeletons/creating-the-skeleton/

Now we can create a label from a model. UI has a new easy way to create such complex labels because your model can provide it out of the box.

**Comment by pauchai** [comment]
What correspondences should be in the code for the attributes from the UI (outside, lock, occluded, hidden)?
skeleton = {
                "confidence": "1", #str(pred_instance["bbox_score"]),
                "label": label["name"],
                "type": "skeleton",
                "elements": [{
                    "label": element["name"],
                    "type": "points",
                    # @NOTICE supose outside attribute  correspondents to presence  in mediapipe
                    "outside": 0 if threshold < pose_landmarks_instance[element["id"]].presence else 1,
                    # @NOTICE supose occluded attriute correspondents visibility in mediapipe
                    "occluded": 0 if threshold < pose_landmarks_instance[element["id"]].visibility else 1,
                    "points": [
                        float(pose_landmarks_instance[element["id"]].x * image.width),
                        float(pose_landmarks_instance[element["id"]].y * image.height)
                    ],
                    "confidence": str(pose_landmarks_instance[element["id"]].presence),
                } for element in label["sublabels"]],
            }

**Comment by bsekachev** [comment]
@pauchai 

``lock, occluded, hidden`` are visualization attributes on UI only. They do not exist on the server. 

If the point is invisbile, outside is True.
Occluded is used when object partially invisible (what is not really applicable for points).

**Comment by bsekachev** [comment]
@pauchai 

> When the point is outside the frame? Is this referred to as "outside"?
When the right leg occludes the left one? Is this considered "occluded"?

I already explained a difference between occluded and outside. 
You know a point coordinates, frame width/height and score from the model. Using these three parameters you easily may decide between occluded or outside.

> And when I don't use points, I believe they shouldn't appear in the UI right sidebar and dont mark as outside.

If you do not need to use these points, do not specify them in skeleton editor in UI constructor.
If you specified them, they are expected to exist in the skeleton, because you setup such configuration for the label.

**Comment by pauchai** [comment]
> If you do not need to use these points, do not specify them in skeleton editor in UI constructor. If you specified them, they are expected to exist in the skeleton, because you setup such configuration for the label.
I m using imported skeleton from model. There is no way to modify it.

File: CHANGELOG.md [hunk 2]
```diff
@@ -16,6 +16,16 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 <!-- scriv-insert-here -->
 
+<a id='changelog-2.8.3'></a>
```
Comment (bsekachev) [inline CHANGELOG.md:19]: Please read in the beginning of CHANGELOG.md how to work with changelog. You do not need to modify the file itself.

File: serverless/tensorflow/mediapipe/nuclio/function.yaml [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+spec:
+  description: Mediapipe body points 33 keypoints
+  runtime: 'python:3.9'
+  handler: main:handler
+  eventTimeout: 30s
+
+  build:
+    image: cvat.tf.mediapipe.33kps
+    baseImage: ubuntu:22.04
```
Comment (bsekachev) [inline serverless/tensorflow/mediapipe/nuclio/function.yaml:62]: Consider using based image like python:3.8 or another one with preinstalled python.

File: serverless/tensorflow/mediapipe/nuclio/function.yaml [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+    directives:
+      preCopy:
+        - kind: RUN
+          value: apt-get update && apt-get -y install curl python3 python3-pip libgl1-mesa-glx python-is-python3 libglib2.0-0
+        - kind: RUN
+          value: pip3 install mediapipe opencv-python numpy matplotlib pyyaml
+        - kind: WORKDIR
+          value: /opt/nuclio
+        - kind: RUN
```
Comment (bsekachev) [inline serverless/tensorflow/mediapipe/nuclio/function.yaml:72]: Consider using ADD command ass in this file https://github.com/opencv/cvat/blob/develop/serverless/pytorch/mmpose/hrnet32/nuclio/function.yaml 

to get rid of extra dependency ``curl``

File: serverless/tensorflow/mediapipe/nuclio/function.yaml [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+          value: curl -O https://storage.googleapis.com/mediapipe-models/pose_landmarker/pose_landmarker_heavy/float16/1/pose_landmarker_heavy.task
+        - kind: RUN
+          value: ln -s /usr/bin/pip3 /usr/local/bin/pip
```
Comment (bsekachev) [inline serverless/tensorflow/mediapipe/nuclio/function.yaml:75]: Is it necessary?

File: serverless/tensorflow/mediapipe/nuclio/main.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    )
+
+    detection_result = context.user_data.model_handler(image)
+    pose_landmarks_instances = detection_result.pose_landmarks
+
+    results = []
+    for pose_landmarks_instance in pose_landmarks_instances:
+        for label in context.user_data.labels:
+            skeleton = {
+                "confidence": "1", #str(pred_instance["bbox_score"]),
```
Comment (bsekachev) [inline serverless/tensorflow/mediapipe/nuclio/main.py:58]: Does not this model provide any score for the whole skeleton?

File: serverless/tensorflow/mediapipe/nuclio/main.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+                "label": label["name"],
+                "type": "skeleton",
+                "elements": [{
+                    "label": element["name"],
+                    "type": "points",
+                    # @NOTICE supose outside atpylint serverless/tensorflow/mediapipe/nuclio/main.pytribute  correspondents to presence  in mediapipe
+                    "outside": 0 if threshold < pose_landmarks_instance[element["id"]].presence else 1,
+                    # @NOTICE supose occluded attriute correspondents visibility in mediapipe
+                    "occluded": 0 if threshold < pose_landmarks_instance[element["id"]].visibility else 1,
+                    #"outside": 0 if threshold < pose_landmarks_instance[element["id"]].visibility else 1,
```
Comment (bsekachev) [inline serverless/tensorflow/mediapipe/nuclio/main.py:68]: Please remove not necessary commented lines

File: serverless/tensorflow/mediapipe/nuclio/main.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+            skeleton = {
+                "confidence": "1", #str(pred_instance["bbox_score"]),
+                "label": label["name"],
+                "type": "skeleton",
+                "elements": [{
+                    "label": element["name"],
+                    "type": "points",
+                    # @NOTICE supose outside atpylint serverless/tensorflow/mediapipe/nuclio/main.pytribute  correspondents to presence  in mediapipe
+                    "outside": 0 if threshold < pose_landmarks_instance[element["id"]].presence else 1,
+                    # @NOTICE supose occluded attriute correspondents visibility in mediapipe
```
Comment (bsekachev) [inline serverless/tensorflow/mediapipe/nuclio/main.py:66]:
```suggestion
                    # suppose occluded attribute corresponds to visibility in mediapipe
```

File: serverless/tensorflow/mediapipe/nuclio/main.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    for pose_landmarks_instance in pose_landmarks_instances:
+        for label in context.user_data.labels:
+            skeleton = {
+                "confidence": "1", #str(pred_instance["bbox_score"]),
+                "label": label["name"],
+                "type": "skeleton",
+                "elements": [{
+                    "label": element["name"],
+                    "type": "points",
+                    # @NOTICE supose outside atpylint serverless/tensorflow/mediapipe/nuclio/main.pytribute  correspondents to presence  in mediapipe
```
Comment (bsekachev) [inline serverless/tensorflow/mediapipe/nuclio/main.py:64]:
```suggestion
                    # suppose outside attribute corresponds to presence in mediapipe
```

File: serverless/tensorflow/mediapipe/nuclio/main.py (hunk 6 above)
Comment (bsekachev) [reply serverless/tensorflow/mediapipe/nuclio/main.py:58]: Since it has parameter ``min_pose_detection_confidence`` I believe it should provide final confidence for the pose

File: serverless/tensorflow/mediapipe/nuclio/main.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    context.logger.info("Init detector...")
+
+    base_options = python.BaseOptions(
+    model_asset_path='pose_landmarker_heavy.task'
+    )
+    options = vision.PoseLandmarkerOptions(
+        base_options=base_options,
+        num_poses = 2,
+        min_pose_detection_confidence = 0.5,
+        min_pose_presence_confidence = 0.5,
```
Comment (bsekachev) [inline serverless/tensorflow/mediapipe/nuclio/main.py:20]: Shall we just initialize them by thershold and remove ``if`` conditions at the end of the file?

File: serverless/tensorflow/mediapipe/nuclio/main.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+import numpy as np
+def init_context(context):
+    context.logger.info("Init detector...")
+
+    base_options = python.BaseOptions(
+    model_asset_path='pose_landmarker_heavy.task'
+    )
+    options = vision.PoseLandmarkerOptions(
+        base_options=base_options,
+        num_poses = 2,
```
Comment (bsekachev) [inline serverless/tensorflow/mediapipe/nuclio/main.py:18]: Why did you strict number of poses by 2?

File: serverless/tensorflow/mediapipe/nuclio/main.py (hunk 10 above)
Comment (bsekachev) [inline serverless/tensorflow/mediapipe/nuclio/main.py:20]: I am not acquainted with mediapipe, but how these parameters used? Why are there 0.5. We have also default threshold 0.55. Can the model return points with presence < 0.5 after detection? 

What if threshold, defined in runtime, after detector initialized less then 0.5, will the detector return these points as not outside?

Comment (pauchai) [reply serverless/tensorflow/mediapipe/nuclio/main.py:20]: That 0.5 value i get from documentation example
**min_pose_detection_confidence**  - The minimum confidence score for the pose detection to be considered successful.
**min_pose_presence_confidence**  - The minimum confidence score of pose presence score in the pose landmark detection.

File: serverless/tensorflow/mediapipe/nuclio/main.py (hunk 11 above)
Comment (pauchai) [reply serverless/tensorflow/mediapipe/nuclio/main.py:18]: i think 0 is unlimited , but i cant find mention about it in docs
Should i set this parameter = 1 in this version?

Comment (bsekachev) [reply serverless/tensorflow/mediapipe/nuclio/main.py:18]: I do not think we want to strict this number somehow

File: serverless/tensorflow/mediapipe/nuclio/main.py (hunk 6 above)
Comment (pauchai) [reply serverless/tensorflow/mediapipe/nuclio/main.py:58]: no only keypoints and segmentation mask is defined in config

File: serverless/tensorflow/mediapipe/nuclio/main.py (hunk 7 above)
Comment (pauchai) [reply serverless/tensorflow/mediapipe/nuclio/main.py]: ok

File: serverless/tensorflow/mediapipe/nuclio/main.py (hunk 8 above)
Comment (pauchai) [reply serverless/tensorflow/mediapipe/nuclio/main.py]: i remain ontly outside:

File: serverless/tensorflow/mediapipe/nuclio/function.yaml (hunk 5 above)
Comment (pauchai) [reply serverless/tensorflow/mediapipe/nuclio/function.yaml:75]: yes it downloads mediapipe model pose_landmarker_heavy.task for detector

Comment (bsekachev) [reply serverless/tensorflow/mediapipe/nuclio/function.yaml:75]: I mean creating a soft link from pip3 to pip

File: serverless/tensorflow/mediapipe/nuclio/function.yaml (hunk 4 above)
Comment (pauchai) [reply serverless/tensorflow/mediapipe/nuclio/function.yaml]: ok

File: serverless/tensorflow/mediapipe/nuclio/function.yaml (hunk 3 above)
Comment (pauchai) [reply serverless/tensorflow/mediapipe/nuclio/function.yaml]: ok

**Comment by bsekachev** [comment]
@pauchai 

From my research, it looks like this model is not able to detect more than one person.. or I am wrong?

**Comment by bsekachev** [comment]
Unfortunately we are not ready to accept the PR as is
It detects only one person per image and I have some doubts about the quality of detecting
